pkg/services/librarypanels/librarypanels.go

```diff
@@ -232,15 +235,85 @@ func connectLibraryPanelsRecursively(c *models.ReqContext, panels []interface{},
 		}
 
 		// we have a library panel
-		uid := libraryPanel.Get("uid").MustString()
-		if len(uid) == 0 {
+		UID := libraryPanel.Get("uid").MustString()
+		if len(UID) == 0 {
 			return errLibraryPanelHeaderUIDMissing
 		}
-		_, exists := libraryPanels[uid]
+		_, exists := libraryPanels[UID]
 		if !exists {
-			libraryPanels[uid] = uid
+			libraryPanels[UID] = UID
 		}
 	}
 
 	return nil
 }
+
+// ImportLibraryPanelsForDashboard loops through all panels in dashboard JSON and creates any missing library panels in the database.
+func (lps *LibraryPanelService) ImportLibraryPanelsForDashboard(c *models.ReqContext, dash *models.Dashboard, folderID int64) error {
+	return importLibraryPanelsRecursively(c, lps.LibraryElementService, dash.Data, folderID)
+}
+
+func importLibraryPanelsRecursively(c *models.ReqContext, service libraryelements.Service, parent *simplejson.Json, folderID int64) error {
+	panels := parent.Get("panels").MustArray()
+	for _, panel := range panels {
+		panelAsJSON := simplejson.NewFromAny(panel)
```

Comment:
probably try to avoid the simplejson package https://github.com/grafana/grafana/blob/main/contribute/style-guides/backend.md, use json-iterator instead.

Reply:
Thanks for the feedback, I've opened a separate issue for this refactor here #39268